Engine/Engine.cs

```diff
@@ -245,11 +245,15 @@ public void Run(AlgorithmNodePacket job, string assemblyPath)
                         // Algorithm runtime error:
                         if (algorithm.RunTimeError != null)
                         {
-                            throw algorithm.RunTimeError;
+                            throw new Exception("Algorithm.RunTimeError", algorithm.RunTimeError); //Keep original exception as inner exception to preserve stack trace
                         }
                     }
                     catch (Exception err)
                     {
+                        while (err != null && err.InnerException != null)
+                        {
+                            err = err.InnerException; //Find inner most exception as this will be most useful to algorithm designer
```

Comment:
Thank you for the PR @robcodes. I think I agree with Stefano the [first part of PR](https://github.com/QuantConnect/Lean/pull/406/commits/ef6801e53afc5c0ff4416229a64b6c9ca789d890#diff-adae0a18f9996ae90d595c4ca363491aR248) is great but if possible can you remove the iterative loop and throw the ToString in the _algorithmHandlers.Results.RuntimeError()/_systemHandlers methods. This should give the user the whole InnerException chain.